doc/python/BACCAB.py

```diff
@@ -1,24 +1,24 @@
-from printer import Format,xpdf
-from ga import Ga
-from mv import Com
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+from galgebra.printer import Format, xpdf
+from galgebra.ga import Ga
 
 Format()
 g4d = Ga('a b c d')
-(a,b,c,d) = g4d.mv()
-
-print 'g_{ij} =',g4d.g
-print '\\bm{a|(b*c)} =',a|(b*c)
-print '\\bm{a|(b^c)} =',a|(b^c)
-print '\\bm{a|(b^c^d)} =',a|(b^c^d)
-print '\\bm{a|(b^c)+c|(a^b)+b|(c^a)} =',(a|(b^c))+(c|(a^b))+(b|(c^a))
-print '\\bm{a*(b^c)-b*(a^c)+c*(a^b)} =',a*(b^c)-b*(a^c)+c*(a^b)
-print '\\bm{a*(b^c^d)-b*(a^c^d)+c*(a^b^d)-d*(a^b^c)} =',\
-            a*(b^c^d)-b*(a^c^d)+c*(a^b^d)-d*(a^b^c)
-print '\\bm{(a^b)|(c^d)} =',(a^b)|(c^d)
-print '\\bm{((a^b)|c)|d} =',((a^b)|c)|d
-print '\\bm{(a^b)\\times (c^d)} =',Com(a^b,c^d)
-xpdf(paper='letter',prog=True)
-
-
-
+(a, b, c, d) = g4d.mv()
 
+print('g_{ij} =', g4d.g)
+print('\\bm{a|(b*c)} =', a | (b * c))
+print('\\bm{a|(b^c)} =', a | (b ^ c))
+print('\\bm{a|(b^c^d)} =', a | (b ^ c ^ d))
+# FIXME this should print 0, got blank
+print('\\bm{a|(b^c)+c|(a^b)+b|(c^a)} =',
+      (a | (b ^ c)) + (c | (a ^ b)) + (b | (c ^ a)))
```

Comment:
Finnally found the reason:

When added together,  `(a | (b ^ c)) + (c | (a ^ b)) + (b | (c ^ a)))` is initially `(a.b)c−(b.c)a−((a.b)c−(b.c)a)`, so it missed the first branch in `GaLatexPrinter._print_Mv`:

```python
    def _print_Mv(self, expr):
        if expr.obj == S(0):   # which equals False at this stage
            return('0 \n')
        else:
            return expr.Mv_latex_str()
```

But in `Mv.Mv_latex_str()`, some code modified `expr.obj`:

```python
        self.obj = expand(self.obj)
        self.characterise_Mv()
        self.obj = metric.Simp.apply(self.obj)
```

After the lines above the self.obj now equals `S(0)`.

So a simple yet hacky fix is to add the following after the code above:

```python
        if self.obj == S(0):
            return ' 0 '
```

But the real fix is to move the simplification-related logic out of the printer method.

The reason was found because I observed:

![image](https://user-images.githubusercontent.com/64258/46296317-ac543480-c5cc-11e8-97c4-789183ed5ecb.png)

which indicated side effect in printing which is caused by:

![image](https://user-images.githubusercontent.com/64258/46296396-de659680-c5cc-11e8-9475-24ce6d59e6dc.png)

which is caused by:

![image](https://user-images.githubusercontent.com/64258/46296434-f4735700-c5cc-11e8-9402-d375d3d011c4.png)

Reply:
> which indicated side effect in printing

Hey look, you saw gh-141 coming over a year ago :)

Reply:
Ah......and finally you nailed it. I must have buried something else like this...